README.md
# MozDef-Triage-Bot
A Slack bot that facilitates triaging MozDef alerts by automating outreach to Mozillians


## Flow

* MozDef : The running instance of the MozDef SIEM
* The Bot : Serverless AWS Lambda Code triggered either by calls to the API Gateway 
  or via direct AWS Lambda invocation with the AWS API
* Slack : The Slack API

### MozDef Triggers the Bot

MozDef Triggers the Bot

1. MozDef, using the dedicated [bot AWS user](https://github.com/mozilla/MozDef-Triage-Bot/blob/master/cloudformation/slack-triage-bot-user.yaml), 
   [invokes The Bot's AWS Lambda function](https://github.com/mozilla/MozDef/blob/bca65c274d363d56c417caada64be05e8585cd68/alerts/actions/triage_bot.py#L497)
   passing The Bot a unique `identifier`, the user to pose the question 
   to, the name of the alert to send the user, the summary text of the alert to 
   send the user and the level of confidence MozDef has in the identity of the user
2. The Bot [sends a direct message to the user on Slack](https://github.com/mozilla/MozDef-Triage-Bot/blob/f36b293c37e407e96a20c3b225ed10467a835d0c/cloudformation/functions/slack_triage_bot_api/app.py#L235)
   with the question via the Slack API

![MozDef Triggers the Bot](https://mermaid.ink/img/eyJjb2RlIjoiZ3JhcGggVERcbiAgICBtb3pkZWZbTW96RGVmXVxuICAgIGJvdFtUaGUgQm90XVxuICAgIHNsYWNrW1NsYWNrIEFQSV1cbiAgICB1c2VyW1VzZXJdXG4gICAgbW96ZGVmIC0tPnxsYW1iZGEuaW52b2tlfCBib3RcbiAgICBib3QgLS0-fFBPU1R8IHNsYWNrXG4gICAgc2xhY2sgLS0-fERpc3BsYXkgbWVzc2FnZXwgdXNlclxuICAgICIsIm1lcm1haWQiOiJ7XG4gIFwidGhlbWVcIjogXCJkZWZhdWx0XCJcbn0iLCJ1cGRhdGVFZGl0b3IiOmZhbHNlLCJhdXRvU3luYyI6dHJ1ZSwidXBkYXRlRGlhZ3JhbSI6ZmFsc2V9)

<details>
  <summary>Click to see the mermaid code</summary>

```mermaid
graph TD
    mozdef[MozDef]
    bot[The Bot]
    slack[Slack API]
    user[User]
    mozdef -->|lambda.invoke| bot
    bot -->|POST| slack
    slack -->|Display message| user
```
</details>

### The User Response

1. The user clicks one of the buttons in the Slack message, indicating their response
2. Slack POSTs to https://myslackbot.example.com/slack/interactive-endpoint
   with the details of the user's response
3. The Bot receives the POST and
   1. [Emits an event to MozDef](https://github.com/mozilla/MozDef-Triage-Bot/blob/f36b293c37e407e96a20c3b225ed10467a835d0c/cloudformation/functions/slack_triage_bot_api/app.py#L344-L351)
      via [an SQS queue created for MozDef to consume](https://github.com/mozilla/MozDef-Triage-Bot/blob/f36b293c37e407e96a20c3b225ed10467a835d0c/cloudformation/functions/slack_triage_bot_api/config.py#L13)
      with a MozDef event with a category of triagebot , the [unique `identifier`](https://github.com/mozilla/MozDef-Triage-Bot/blob/f36b293c37e407e96a20c3b225ed10467a835d0c/cloudformation/functions/slack_triage_bot_api/app.py#L345)
      that MozDef [created above](https://github.com/mozilla/MozDef-Triage-Bot/blob/f36b293c37e407e96a20c3b225ed10467a835d0c/cloudformation/functions/slack_triage_bot_api/app.py#L460)
      when it triggered the bot, the identity of the user and the user's response
   2. [POSTs back](https://github.com/mozilla/MozDef-Triage-Bot/blob/f36b293c37e407e96a20c3b225ed10467a835d0c/cloudformation/functions/slack_triage_bot_api/app.py#L358-L361)
      to Slack to update the Slack UI to show that The Bot received the user's 
      button click. This adds the `Understood, thanks for letting us know.` annotation
      to the Slack message.

![The User Response](https://mermaid.ink/img/eyJjb2RlIjoiZ3JhcGggVERcbiAgICBtb3pkZWZbTW96RGVmXVxuICAgIGJvdFtUaGUgQm90XVxuICAgIHNsYWNrW1NsYWNrIEFQSV1cbiAgICB1c2VyW1VzZXJdXG4gICAgYXBpZ2F0ZXdheVtteXNsYWNrYm90LmV4YW1wbGUuY29tXVxuICAgIHNxc1tNb3pEZWYgU1FTIFF1ZXVlXVxuICAgIHVzZXIgLS0-fENsaWNrIG1lc3NhZ2UgYnV0dG9ufCBzbGFja1xuICAgIHNsYWNrIC0tPnxQT1NUfCBhcGlnYXRld2F5XG4gICAgYXBpZ2F0ZXdheSAtLT58aW52b2tlfCBib3RcbiAgICBib3QgLS0-fHNxcy5zZW5kX21lc3NhZ2V8IHNxc1xuICAgIG1vemRlZiAtLT58c3FzLnJlY2VpdmVfbWVzc2FnZXwgc3FzIiwibWVybWFpZCI6IntcbiAgXCJ0aGVtZVwiOiBcImRlZmF1bHRcIlxufSIsInVwZGF0ZUVkaXRvciI6ZmFsc2UsImF1dG9TeW5jIjp0cnVlLCJ1cGRhdGVEaWFncmFtIjpmYWxzZX0)

<details>
  <summary>Click to see the mermaid code</summary>

```mermaid
graph TD
    mozdef[MozDef]
    bot[The Bot]
    slack[Slack API]
    user[User]
    apigateway[myslackbot.example.com]
    sqs[MozDef SQS Queue]
    user -->|Click message button| slack
    slack -->|POST| apigateway
    apigateway -->|invoke| bot
    bot -->|sqs.send_message| sqs
    mozdef -->|sqs.receive_message| sqs
```
</details>

## Deployment

In these instructions we'll assume that the domain name that the bot is deployed
to is `myslackbot.example.com`. You can find the actual domain name of the API in the
`PROD_DOMAIN_NAME` or `DEV_DOMAIN_NAME` variables in the [`Makefile`](Makefile). 

* [Create the Slack App](https://api.slack.com/apps?new_app=1)
  * Go to the `Collaborators` section of the configuration and add all the users
    that will administer the bot.
  * Go to the `App Home` section
    * In the `Your App’s Presence in Slack` area click `Edit` next to `App Display Name`
      and set the `Display Name (Bot Name)` to `mozilla_security_bot` and the `Default Name`
      to `mozilla_security_bot` 
  * Go to the `Interactivity & Shortcuts` section
    * In the `Interactivity` area set the `Request URL` to `https://myslackbot.example.com/slack/interactive-endpoint`
    * In the `Select Menus` area set the `Options Load URL` to `https://myslackbot.example.com//slack/options-load-endpoint`
  * Go to the `OAuth & Permissions` section of the configuration
    * Ignore the `Bot User OAuth Access Token` as we don't use it
    * Update the `Redirect URLs` to include the `https://myslackbot.example.com/redirect_uri`
      URL.
  * Scopes
    * The Bot Token Scopes needed are
      * `users:read.email` : https://api.slack.com/methods/users.lookupByEmail
      * `users:read` : This is required because of users:read.email
      * `im:write` : https://api.slack.com/methods/conversations.open
      * `chat:write` : https://api.slack.com/methods/chat.postMessage
  * Install the App
* Create the AWS IAM user that MozDef will use to interact with the Lambda function
  and SQS queue
  `make deploy-mozdef-slack-triage-bot-user`
* Determine the Slack Client Secret
  * This can be done by going to the Slack App's configuration (at a URL like
    `https://api.slack.com/apps/ABCDE123AB/general`).
  * Navigate to https://api.slack.com/apps
  * Click your app to get to the configuration page
  * Find the `Client Secret` in the `App Credentials` section
* Run the make command for the environment you want
    ```shell script
    PROD_SLACK_CLIENT_SECRET=0123456789abcdef0123456789abcdef make deploy-mozdef-slack-triage-bot-api
    ```
    
    or
    
    ```shell script
    DEV_SLACK_CLIENT_SECRET=0123456789abcdef0123456789abcdef make deploy-mozdef-slack-triage-bot-api-dev
    ```
    
    depending on the account
* Provision a Slack OAuth token for the bot by browsing to the 
  `https://myslackbot.example.com/authorize` URL of the deployed API in any 
  browser. 
  By hitting this `/authorize` URL, it will cause the API to contact Slack and
  request an OAuth token which the bot will then store in AWS SSM Parameter Store.
  This step need only be done once as, from this point on, the bot will continue
  to use this token which is stored in the Parameter Store.

## Testing

You can test invoking the function by passing the email address of the user
you want to send a message to and calling

```shell script
EMAIL_ADDRESS=user@example.com make test-mozdef-slack-triage-bot-api-invoke
```

which will pass

```json
{
    "identifier":"9Zo02m4B7gIfixq3c4Xh",
    "alert":"duo_bypass_codes_generated",
    "identityConfidence":"lowest",
    "summary":"DUO bypass codes have been generated for your account. ",
    "user":"user@example.com"
}
```

to the API which will return the JSON response from Slack of the message that
was sent to the user.

You can also test the API Gateway interface by running

```shell script
make test-mozdef-slack-triage-bot-api-http
```

which will hit the `/test` API endpoint and get back a 200 `API request received`

You can also visit the `/error` endpoint to get a 400 or any other endpoint to get a 404

## Discovering the SQS URL containing user responses

To discover the URL of the SQS queue into which user responses are sent, call

```shell script
make discover-sqs-queue-url 
```

which will return a value like

```json
{"result": "https://sqs.us-west-2.amazonaws.com/012345678901/MozDefSlackTriageBotAPI-SlackTriageBotMozDefQueue-ABCDEFGHIJKL"}
```

## Discovering the Lambda function name

Call the [lambda:ListFunctions](https://docs.aws.amazon.com/lambda/latest/dg/API_ListFunctions.html)
API and filter the results based on the name. You can see an example of this by
running the make command

```shell script
make discover-lambda-function-name 
```

which will return a value like `MozDefSlackTriageBotAPI-SlackTriageBotApiFunction-1N9KLDX1926F3`

## Fetching User API Keys

You can fetch the User API keys from the CloudFormation outputs

```shell script
make show-user-credentials
```

## Integrating MozDef and the MozDef-Triage-Bot

The CloudFormation templates in this repo provision an AWS IAM user as well as
the SQS queue which MozDef receives messages from.

To integrate MozDef with the bot
* Deploy the `slack-triage-bot-user.yaml` CloudFormation template
* Deploy the `slack-triage-bot-api.yaml` CloudFormation template
* Gather the 3 sets of information you'll need to put into the `MozDef-deploy`
  Ansible repo to configure MozDef
  * The API keys 
    * either look at the stack outputs in the user stack or run 
      `make show-user-credentials`
  * The SQS Queue name, AWS region and AWS account ID
    * either look at the stack outputs in the api stack or run 
      `make discover-sqs-queue-url`
  * The Lambda function name
     * either look at the stack outputs in the api stack or run 
       `make discover-lambda-function-name`
  
The API keys will grant MozDef permission to both invoke the lambda function
as well as receive messages from the SQS queue

## Mozilla's deployments in Slack

* [Mozilla deployment](https://api.slack.com/apps/AS6G90NUT/general)
* [mozilla-sandbox-SCIM deployment](https://api.slack.com/apps/AR6G404SH/general)

More Mozilla specific information can be found in mana : https://mana.mozilla.org/wiki/display/SECURITY/Mozilla+Security+Bot+or+MozDef+Triage+Bot


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

### Added

* Diagrams and flow descriptions to the README [#26](https://github.com/mozilla/MozDef-Triage-Bot/pull/26) [#27](https://github.com/mozilla/MozDef-Triage-Bot/pull/27)

## [1.2.0] - 2020-04-20

### Changed

* Makefile by adding check for missing client_secrets [#18](https://github.com/mozilla/MozDef-Triage-Bot/pull/18)
* Deployment scripts to support deploying from MacOS [#20](https://github.com/mozilla/MozDef-Triage-Bot/pull/20)

### Fixed

* store_oauth_token ValidationException error [#21](https://github.com/mozilla/MozDef-Triage-Bot/pull/21)
* typo in IAM permission for SSM [#22](https://github.com/mozilla/MozDef-Triage-Bot/pull/22) [#24](https://github.com/mozilla/MozDef-Triage-Bot/pull/24)

## [1.1.0] - 2020-01-02

### Changed

* Exception handling [#9](https://github.com/mozilla/MozDef-Triage-Bot/pull/9) [#10](https://github.com/mozilla/MozDef-Triage-Bot/pull/10)
  * Deal with exceptions differently depending on if the function was called
    by API Gateway or through direct invocation.
  * Pass SlackExceptions through to the invoker of the function.
* IAM user creation, moving it into its own CloudFormation stack [#11](https://github.com/mozilla/MozDef-Triage-Bot/pull/11)
  * This will make the IAM user and API keys durable across API
    stack rebuilds.
    This also grants the IAM user rights to invoke all deployments
    of the API and all SQS queues used by the API
* The result structure one more time to be a JSON dictionary [#11](https://github.com/mozilla/MozDef-Triage-Bot/pull/11)
  * This was changed in 5d0ba05 and then in 74f8d5f
    This is another attempt at getting it right

### Added

* Add support for emitting slackName to MozDef
  * This emits an additional piece of information to MozDef about
    the user, the user's Slack username in the slackName field of
    the SQS message

## [1.0.0] - 2019-12-27

Big update to the bot [#5](https://github.com/mozilla/MozDef-Triage-Bot/pull/5)

### Added

* Functions to utils
    * store_oauth_token : Store an OAuth 2 access token in SSM parameter store
    * get_access_token : Fetch the OAuth 2 access token for a given client_id from cache or SSM
      parameter store
    * emit_to_mozdef : Send a message with the user's response to SQS for pickup by MozDef
    * call_slack : POST to a slack URL and return the result
    * provision_token : Given an OAuth 2 code, obtain a Slack access token and store it
    * redirect_to_slack_authorize : Build a Slack OAuth 2 authorization URL and redirect the user to it
* Functions to app
    * API calls
        * process_api_call : Process an API Gateway call depending on the URL path called
            * `/authorize` : utils.redirect_to_slack_authorize
            * `/redirect_uri` : utils.provision_token
            * `/slack/interactive-endpoint` : handle_message_interaction : Process a user's interaction with a Slack message
                * send_slack_message_response : Respond to a user's selection by updating the Slack message with a reply
    * Direct invocations
        * send_message_to_slack : Send a message to a user via IM or Slack App conversation
            * get_user_from_email : Fetch a slack user dictionary for an email address
            * compose_message : Create a Slack message object
            * create_slack_channel : Create an IM channel with a user
            * post_message : Post a message to a slack channel
* Details on discovering the SQS URL to the README
* Additional test invocations to the Makefile
* SQS URL discovery to the Makefile
* `requests` to the requirements.txt
* New settings to config.py
* A function invoker IAM user with access keys for MozDef to use
  Add working end to end bot [#2](https://github.com/mozilla/MozDef-Triage-Bot/pull/2)

### Changed

* lambda_handler in app to cover all URL paths and direct invocations
* README configuration details
* CloudFormation template to
    * accept Slack client ID and secret
    * Grant Lambda function rights to
        * read and write to SSM parameter store
        * decrypt parameter store secrets
        * send messages to the SQS queue
    * Create the SQS queue
    * Grant the MozDef user rights to read from the SQS queue
    
## [0.9.0] - 2019-11-26

### Added

- The initial slack bot code [#1](https://github.com/mozilla/MozDef-Triage-Bot/pull/1)

## [0.1.0] - 2019-11-01

### Added

- The initial design ideas and specifications for the slack bot

[unreleased]: https://github.com/mozilla/MozDef-Triage-Bot/compare/v1.2.0...HEAD
[1.2.0]: https://github.com/mozilla/MozDef-Triage-Bot/compare/v1.1.0...v1.2.0
[1.1.0]: https://github.com/mozilla/MozDef-Triage-Bot/compare/v1.0.0...v1.1.0
[1.0.0]: https://github.com/mozilla/MozDef-Triage-Bot/compare/v0.9.0...v1.0.0
[0.9.0]: https://github.com/mozilla/MozDef-Triage-Bot/compare/v0.1.0...v0.9.0
[0.1.0]: https://github.com/mozilla/MozDef-Triage-Bot/releases/tag/v0.1.0


tests/test_mozdef_alert_triage_bot.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

"""Tests for `mozdef_alert_triage_bot` package."""

import pytest


from mozdef_alert_triage_bot import mozdef_alert_triage_bot


@pytest.fixture
def response():
    """Sample pytest fixture.

    See more at: http://doc.pytest.org/en/latest/fixture.html
    """
    # import requests
    # return requests.get('https://github.com/audreyr/cookiecutter-pypackage')


def test_content(response):
    """Sample pytest test function with the pytest fixture as an argument."""
    # from bs4 import BeautifulSoup
    # assert 'GitHub' in BeautifulSoup(response.content).title.string


cloudformation/functions/slack_triage_bot_api/app.py
import json
import logging
import traceback
import urllib.parse
import requests

from .config import CONFIG

from .utils import (
    call_slack,
    emit_to_mozdef,
    provision_token,
    redirect_to_slack_authorize,
    SlackException
)

logger = logging.getLogger()
logging.getLogger().setLevel(CONFIG.log_level)
logging.getLogger('boto3').propagate = False
logging.getLogger('botocore').propagate = False
logging.getLogger('urllib3').propagate = False


def get_user_from_email(email: str) -> dict:
    """Fetch a slack user dictionary for an email address

    Required slack scopes
    * bot - users:read.email : https://api.slack.com/methods/users.lookupByEmail
    * bot - users:read : This scope must be requested if users:read.email is
                         requested
    :param email: email address of the slack user
    :return: dictionary of user information
    """
    data = {'email': email}
    url = 'https://slack.com/api/users.lookupByEmail'
    return call_slack(url, data, 'user')


def create_slack_channel(user: str) -> dict:
    """Create an IM channel with a user

    Required slack scopes
    * bot - im:write : https://api.slack.com/methods/conversations.open

    :param user: The Slack user ID of the user to create an IM channel with
    :return: dictionary of channel information
    """
    data = {'users': user}
    url = 'https://slack.com/api/conversations.open'
    return call_slack(url, data, 'channel', True)


def compose_message(
        identifier: str,
        alert: str,
        summary: str,
        email: str,
        user: dict,
        identity_confidence: str) -> dict:
    """Create a Slack message object

    :param identifier: The unique identifier for this message
    :param alert: The name of the MozDef alert
    :param summary: The summary text of the alert
    :param email: The email address of the user
    :param user: The slack user dictionary
    :param identity_confidence: The identity confidence sent from MozDef
    :return: A Slack message dictionary
    """

    default_response = {
        'identifier': identifier,
        'email': email,
        'slack_name': user['name'],
        'alert': alert,
        'identity_confidence': identity_confidence
    }

    # TODO : Add something that if the identity_confidence is high don't offer
    # the wronguser option
    blocks = [
        {
            "block_id": "mozdef-triage-bot-api-question",
            "text": {
                "text": "{}\nWas this action taken by you ({})?".format(
                    summary, email),
                "type": "mrkdwn",
            },
            "type": "section"
        },
        {
            "block_id": "mozdef-triage-bot-api-answer",
            "type": "actions",
            "elements": [
                {
                    "action_id": "mozdef-triage-bot-api-yes",
                    "style": "primary",
                    "text": {
                        "emoji": False,
                        "text": "Yes, I did that",
                        "type": "plain_text"
                    },
                    "type": "button",
                    "value": json.dumps(
                        {**default_response, "response": "yes"})
                },
                {
                    "action_id": "mozdef-triage-bot-api-no",
                    "style": "danger",
                    "text": {
                        "emoji": False,
                        "text": "No, I didn't do that!",
                        "type": "plain_text",
                    },
                    "type": "button",
                    "confirm": {
                        "confirm": {
                            "text": "Ya, I didn't take that action",
                            "type": "plain_text"
                        },
                        "deny": {
                            "text": "Oh, nevermind, I did do that",
                            "type": "plain_text"
                        },
                        "text": {
                            "text": "Are you sure that you didn't take that "
                                    "action? If you're sure then someone in "
                                    "the security team will contact you to "
                                    "follow up.".format(email),
                            "type": "mrkdwn"
                        },
                        "title": {
                            "text": "Are you sure?", "type": "plain_text"}
                    },

                    "value": json.dumps({**default_response, "response": "no"})
                }
            ]
        }
    ]
    if identity_confidence in ['moderate', 'low', 'lowest']:
        blocks[1]['elements'].append(
            {
                "action_id": "mozdef-triage-bot-api-wronguser",
                "text": {
                    "text": "You've got the wrong person",
                    "type": "plain_text"
                },
                "confirm": {
                    "confirm": {
                        "text": "Ya, that's not me",
                        "type": "plain_text"
                    },
                    "deny": {
                        "text": "Oh, actually that is me",
                        "type": "plain_text"
                    },
                    "text": {
                        "text": (
                            "Are you sure that {} isn't you and we've sent "
                            "this alert to the wrong user?".format(email)),
                        "type": "mrkdwn"
                    },
                    "title": {"text": "Are you sure?", "type": "plain_text"}
                },
                "type": "button",
                "value": json.dumps(
                    {**default_response, "response": "wronguser"})
            }
        )
    blocks[1]['elements'].append(
        {
            "action_id": "mozdef-triage-bot-api-notsure",
            "text": {
                "text": "Hmm... I'm not sure",
                "type": "plain_text"
            },
            "type": "button",
            "value": json.dumps({**default_response, "response": "notsure"})
        }
    )
    blocks_json = json.dumps(blocks)

    # We can't pass "as_user": False here as doing so causes the Slack API to
    # return an error that the chat:write:bot scope is missing
    # Slack support reports this is a known bug
    # https://mozilla-sandbox-scim.slack.com/help/requests/2564183
    message = {
        'blocks': blocks_json,
        'text': summary
    }
    return message


def post_message(channel: str, message: dict) -> dict:
    """Post a message to a slack channel

    Required slack scopes
    * bot - chat:write : https://api.slack.com/methods/chat.postMessage

    :param channel: The Slack channel ID to post the message to
    :param message: The message to post
    :return: A slack message dictionary
    """
    data = message
    data['channel'] = channel
    url = 'https://slack.com/api/chat.postMessage'
    return call_slack(url, data, 'message', True)


def send_message_to_slack(
        identifier: str,
        alert: str,
        summary: str,
        email_address: str,
        identity_confidence: str) -> dict:
    """Send a message to a user via IM or Slack App conversation

    :param identifier: The unique identifier sent by MozDef originally
    :param alert: The name of the MozDef alert
    :param summary: The summary text of the alert
    :param email_address: The user's email address
    :param identity_confidence: The identityConfidence sent by MozDef
                                originally
    :return: A slack message dictionary
    """
    send_to_im = False
    user = get_user_from_email(email_address)
    message = compose_message(
        identifier, alert, summary, email_address, user, identity_confidence)
    if send_to_im:
        channel = create_slack_channel(user['id'])
        post_result = post_message(channel['id'], message)
    else:
        post_result = post_message(user['id'], message)
    return post_result


def send_slack_message_response(
        response_url: str,
        message: dict,
        user_response: str) -> bool:
    """Respond to a user's selection by updating the Slack message with a reply

    :param response_url: The Slack URL to send message responses to
    :param message: The original message that the user made a selection from
    :param user_response: The user's selection
    :return: Whether or not the response to the user succeeded
    """
    if user_response == 'yes':
        bot_response = (
            ':heavy_check_mark: Understood, thanks for letting us know.')
    elif user_response == 'no':
        bot_response = (
            ':open_mouth: Got it, thank you. Someone from the security team '
            'will contact you to follow up on this.')
    elif user_response == 'wronguser':
        bot_response = (
            ":flushed: Oh, sorry about that. Someone from the security team "
            "will look into this and contact the right user. Sorry to bother "
            "you.")
    elif user_response == 'notsure':
        bot_response = (
            ":ok_hand: No problem. Someone from the security team will "
            "contact you to follow up on this.")
    else:
        bot_response = (
            ":heavy_multiplication_x: Hmm, I had some kind of internal error. "
            "Would you contact the security team to let them know that I'm "
            "unwell?")

    if 'mozdef-triage-bot-api-response' in [x.get('block_id') for x
                                            in message.get('blocks', [])]:
        bot_response = "You've changed your mind, no problem. " + bot_response
    response_block = {
        'block_id': 'mozdef-triage-bot-api-response',
        'text': {
            'text': bot_response,
            'type': "mrkdwn"
        },
        "type": "section"
    }
    if 'blocks' in message:
        for i in range(0, len(message['blocks'])):
            if message['blocks'][i].get('block_id') == 'mozdef-triage-bot-api-response':
                message['blocks'][i] = response_block
                break
        else:
            message['blocks'].append(response_block)

    message['replace_original'] = True
    try:
        response = requests.post(
            url=response_url,
            json=message
        )
        response.raise_for_status()
    except requests.exceptions.RequestException as e:
        logger.error(
            'POST of response to {} failed {} : {} : {} : {}'.format(
                response_url,
                message,
                e,
                getattr(e.response, 'status_code', None),
                getattr(e.response, 'text', None)
            ))
        return False
    return True


def handle_message_interaction(payload: dict) -> bool:
    """Process a user's interaction with a Slack message

    payload['type'] :
        'block_actions' : Parse the value that the user chose, send it to
                          MozDef and send a response to the user
    :param payload: A dictionary of data sent from Slack about a user's
                    interaction
    :return: Whether or not the response to the user succeeded
    """
    # The right way to do this is
    # 1. Drop a message in a message queue (e.g. SQS) with the message to
    #    send back to the user
    # 2. Return 200
    # 3. Pull that message off the queue
    # 4. POST to the response_url with the response message
    # Until that's added we'll just
    # 1. POST to the response_url with the response message
    # 2. Hope that the POST completes in under 3 seconds and return 200
    if payload.get('type') == 'block_actions':
        # User clicked a Block Kit interactive component
        for action in payload.get('actions', []):
            if 'value' not in action:
                raise SlackException(
                    'Action encountered with no value : {}'.format(action))
            try:
                value = json.loads(action['value'])
            except json.decoder.JSONDecodeError as e:
                logger.error('Failed to parse button value "{}" : {}'.format(
                    action['value'],
                    e
                ))
                raise
            message_id = emit_to_mozdef(
                value.get('identifier'),
                value.get('email'),
                payload.get('user', {}).get('id'),
                value.get('slack_name'),
                value['identity_confidence'],
                value.get('response')
            )
            allowed_keys = [
                'text', 'blocks', 'attachments', 'thread_ts', 'mrkdwn']
            original_message = {
                k: v for k, v in
                payload.get('message', {}).items()
                if k in allowed_keys}
            return send_slack_message_response(
                payload.get('response_url'),
                original_message,
                value.get('response'))
    else:
        # https://api.slack.com/interactivity/handling#payloads
        logger.error(
            "Encountered a message interaction payload type that hasn't yet "
            "been developed : {}".format(payload))
        return False


def process_api_call(
        event: dict,
        query_string_parameters: dict,
        body: dict) -> dict:
    """Process an API Gateway call depending on the URL path called

    :param event: The API Gateway request event
    :param query_string_parameters: A dictionary of query string parameters
    :param body: The parsed body that was POSTed to the API Gateway
    :return: A dictionary of an API Gateway HTTP response
    """
    if event.get('path') == '/error':
        return {
            'headers': {'Content-Type': 'text/html'},
            'statusCode': 400,
            'body': "Since you requested the /error API endpoint I'll go "
                    "ahead and serve back a 400"}
    elif event.get('path') == '/test':
        return {
            'headers': {'Content-Type': 'text/html'},
            'statusCode': 200,
            'body': 'API request received'}
    elif event.get('path') == '/redirect_uri':
        return provision_token(query_string_parameters)
    elif event.get('path') == '/authorize':
        return redirect_to_slack_authorize()
    elif event.get('path') == '/slack/interactive-endpoint':
        for payload_raw in body.get('payload', []):
            payload = json.loads(payload_raw)
            logger.debug('payload is {}'.format(payload))
            result = handle_message_interaction(payload)
        return {
            'headers': {'Content-Type': 'text/html'},
            'statusCode': 200,
            'body': 'Acknowledged'}
    elif event.get('path') == '/slack/options-load-endpoint':
        # https://api.slack.com/reference/block-kit/block-elements#external_select
        return {
            'headers': {'Content-Type': 'text/html'},
            'statusCode': 200,
            'body': 'Acknowledged'}
    else:
        return {
            'headers': {'Content-Type': 'text/html'},
            'statusCode': 404,
            'body': "That path wasn't found"}


def lambda_handler(event: dict, context: dict) -> dict:
    """Handler for all API Gateway requests

    :param event: AWS API Gateway input fields for AWS Lambda
    :param context: Lambda context about the invocation and environment
    :return: An AWS API Gateway output dictionary for proxy mode
    """
    logger.debug('event is {}'.format(event))
    if event.get('resource') == '/{proxy+}':
        try:
            headers = event['headers'] if event['headers'] is not None else {}
            cookie_header = headers.get('Cookie', '')
            referer = headers.get('Referer', '')
            query_string_parameters = (
                event['queryStringParameters']
                if event['queryStringParameters'] is not None else {})

            parser = None
            if headers.get('Content-Type') == 'application/x-www-form-urlencoded':
                parser = urllib.parse.parse_qs
            elif headers.get('Content-Type') == 'application/json':
                parser = json.loads
            if parser is not None and event.get('body') is not None:
                body = parser(event['body'])
            else:
                body = {}
            return process_api_call(event, query_string_parameters, body)
        except Exception as e:
            logger.error(str(e))
            logger.error(traceback.format_exc())
            return {
                'headers': {'Content-Type': 'text/html'},
                'statusCode': 500,
                'body': 'Error'}
    else:
        # Not an API Gateway invocation, we'll assume a direct Lambda invocation
        try:
            if event.get('action') == 'discover-sqs-queue-url':
                result = {"result": CONFIG.queue_url}
            else:
                try:
                    result = send_message_to_slack(
                        event.get('identifier'),
                        event.get('alert'),
                        event.get('summary'),
                        event.get('user'),
                        event.get('identityConfidence')
                    )
                except SlackException as e:
                    result = {"result": e}
        except Exception as e:
            result = {"result": str(e)}
        return result